### ISSUE TITLE
AWS TransferManager Transfer completes but isDone() still returns false for quite some time.

### ISSUE THREAD

**Issue by EXPEpchaurasia** [aws/aws-sdk-java#1277]
I am using AWS [TransferManager](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/TransferManager.java) for managing large file downloads from S3 in my application. I am trying to print completed transfer percentage every 10 seconds until completed:

```
private void printTransferProgressState(final Transfer transfer) throws Exception {
        DecimalFormat progressPercentFormat = new DecimalFormat("#.##");
        while (!transfer.isDone()) {
            LOGGER.info("Transfer {}% Done.", progressPercentFormat.format(transfer.getProgress().getPercentTransferred()));
            sleep(10 * MILLIS_PER_SECOND);
        }
    }
```
Although it prints the progress report fine, but seems like the [transfer.isDone()](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/Transfer.java#L62) keeps returning **false** for quite sometime even when the transfer is completed 100%. This is the log of download :

```
07:29:23 level=INFO message=Transfer 0% Done.
07:29:33 level=INFO message=Transfer 30.65% Done.
07:29:43 level=INFO message=Transfer 53.5% Done.
07:29:53 level=INFO message=Transfer 90.28% Done.
07:30:03 level=INFO message=Transfer 100% Done.
07:30:13 level=INFO message=Transfer 100% Done.
07:30:23 level=INFO message=Transfer 100% Done.
07:30:33 level=INFO message=Transfer 100% Done.
07:30:43 level=INFO message=Transfer 100% Done.
```

If you see, the transfer was completed 100% by 07:30:03 itself, but `transfer.isDone()` is still returning **false** for next ~40 seconds.
To be specific, it was a [Download](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/Download.java).

I am using sdk version :  **_com.amazonaws:aws-java-sdk-s3:jar:1.11.22_**

I've also mentioned the issue is [stack-overflow](https://stackoverflow.com/questions/45630245/aws-transfermanager-transfer-completes-but-isdone-still-returns-false-for-some)

**Comment by varunnvs92** [comment]
I see that you are using version 1.11.22. There are some significant bugs in early versions of 1.11.x series. Please upgrade to the latest version of the SDK (currently 1.11.180). FYI, there are no breaking changes across versions in 1.11.x series.

I don't think this issue is related to above mentioned bugs. I will investigate this. Can you also try the same code with latest version and see if you can reproduce it?

**Comment by varunnvs92** [comment]
I am guessing you seeing this when downloading multipart objects. This is the expected behavior.  [This](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/DownloadCallable.java#L174-L197) is the code to download objects in parts. The parts are downloaded at line 191 which is where the transfer progress is updated. After all parts are downloaded (100% of data is transferred), all the part files will be merged into a single file. Related code [here](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CompleteMultipartDownload.java#L45-L51). So only after the parts are merged, the transfer state is set to Completed which makes isDone() return true. 

The delay you are seeing is the time between downloading all parts and merging them into single file. I am surprised though that it takes 40 sec to merge all individual parts. I tested with 5GB object and merge took 5 sec for me. As I mentioned in last message, please upgrade to latest version and give it a try. If the wait time is still long, feel free to reopen the issue. Thank you.

**Comment by EXPEpchaurasia** [comment]
@varunnvs92 Thanks for the information. I tried with updated version of AWS SDK as you mentioned. I tried it 3 times so far and i have'nt encountered this case anymore so far. Just for the info, i am trying to download a file that is 1GB in size.

Thanks a lot.

**Comment by EXPEpchaurasia** [comment]
@varunnvs92  Could we please reopen this one. I encountered it happenning again. This time i [disabledParallelDownloading](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/transfer/TransferManagerBuilder.html#disableParallelDownloads--) in TransferManager and its still stuck on 100% for like a minute. This time, there wasn't even anything to merge since it downloaded the entire 1GB file as a single chunk (I verified this).

```
18:19:04 logger=AWSHelper message=Starting Download.
18:19:04 logger=AWSHelper message=Transfer 0% Done.
18:20:54 logger=AWSHelper message=Transfer 100% Done.
18:20:44 logger=AWSHelper message=Transfer 100% Done.
18:20:34 logger=AWSHelper message=Transfer 100% Done.
18:20:24 logger=AWSHelper message=Transfer 100% Done.
18:20:14 logger=AWSHelper message=Transfer 100% Done.
18:20:04 logger=AWSHelper message=Transfer 100% Done.
18:19:54 logger=AWSHelper message=Transfer 100% Done.
18:19:44 logger=AWSHelper message=Transfer 100% Done.
18:19:34 logger=AWSHelper message=Transfer 100% Done.
18:19:24 logger=AWSHelper message=Transfer 100% Done.
18:21:04 logger=AWSHelper message=Download complete. Total time taken : 120 seconds.
```

Could you please suggest ?

**Comment by varunnvs92** [comment]
Can you show the code on how you initialized transferManager and disabled parallel downloads?

**Comment by EXPEpchaurasia** [comment]
Yes sure Varun,

```
    private final AmazonS3 m_s3Client;
    private final TransferManager transferManager;

    public AWSHelper(final String accessKey, final String secretKey) {
        AWSCredentials m_credentials = new BasicAWSCredentials(accessKey, secretKey);

        ClientConfiguration clientConfiguration = new ClientConfiguration();
        clientConfiguration.setMaxErrorRetry(3);
        
String region = Regions.getCurrentRegion() != null ? Regions.getCurrentRegion().getName() : Regions.US_WEST_2.getName();
        
AWSCredentialsProvider credentialsProvider = new AWSStaticCredentialsProvider(m_credentials);
        m_s3Client = AmazonS3ClientBuilder.standard()
                             .withClientConfiguration(clientConfiguration)
                             .withCredentials(credentialsProvider)
                             .withRegion(region)
                             .withForceGlobalBucketAccessEnabled(Boolean.TRUE)
                             .build();

        transferManager = TransferManagerBuilder.standard()
                                  .withS3Client(m_s3Client)
                                  .withMinimumUploadPartSize(MIN_PART_SIZE)
                                  .withDisableParallelDownloads(Boolean.TRUE)
                                  .build();
    }
```

**Comment by EXPEpchaurasia** [comment]
This is happening in AWS EC2 boxes of type c4.4xlarge. The write medium is pretty fast since it writes to the same medium while downloading part files and storing (i verified this). Its the merging of part files and something unknown (if exists) after that which is causing this huge delay. FYI, the total file size is 1GB while the part size is set to 50MB.

**Comment by dagnir** [comment]
Gotcha. Thanks for the additional information! I'll try and repro this and see what I can find.

**Comment by EXPEpchaurasia** [comment]
Any leads here ?

**Comment by varunnvs92** [comment]
Are you seeing the issue for both serial and parallel downloads? In the beginning of the issue, you mentioned that the issue is not present for parallel downloads anymore after upgrading to latest SDK.

**Comment by EXPEpchaurasia** [comment]
That was in a few retries initially, but its intermittently freezing in both serial and parallel downloads. Upgrading to latest SDK didn't help.

**Comment by varunnvs92** [comment]
Sorry for the silence on the ticket. I was not able to reproduce it for serial downloads. For parallel downloads, I was able to reproduce only once for around 10 tries and it was a transient issue and the download was retried again. Can you provide us a complete code that can reproduce this issue?
Please provide information about your system configuration too.

**Comment by millems** [comment]
Closing for now. Please let us know if you are able to reproduce this consistently! It's a hard one to nail down.
